### PR TITLE
ci: Rework the release process

### PR DESCRIPTION
**Pull request by michalborkowski96**
#### Summary
References #4982

@adriansmares I guess I'd like thumbs up from you before I ask others.

#### Changes
Reworked the release flow. The new flow would be:

1. Run github action that creates a `release/v3.x.y` branch
2. Run github action that creates a release candidate
3. Run github action that actually makes the release

For details, refer to updated issue template.

#### Testing
The new github actions have been tested. The thing that has not been tested is the goreleaser itself, but I assume that since changes there were minimal, it should work.

##### Regressions
None

#### Checklist

- [ ] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [ ] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/release.md
Comment (adriansmares) [inline .github/ISSUE_TEMPLATE/release.md]: Probably put the name here directly, as the file name is not visible in the UI. Same below.

File: .github/ISSUE_TEMPLATE/release.md
Comment (adriansmares) [inline .github/ISSUE_TEMPLATE/release.md:25]: Is it not possible to tag via an action too ? Then the tagging action can also open a PR that merges the version increment in the original branch.

File: .goreleaser.release.yml
Comment (adriansmares) [inline .goreleaser.release.yml]: I wonder why the `latest` tag wasn't automatically pushed before. @htdvisser do you recall by any chance ?

File: .github/ISSUE_TEMPLATE/release.md
Comment (michalborkowski96) [reply .github/ISSUE_TEMPLATE/release.md:25]: > Is it not possible to tag via an action too ?

This can be done, but we need some human to push the button "release this now" anyway, and whether it's via github.com, github CLI, git tag, or whatever, doesn't make any difference.

> Then the tagging action can also open a PR that merges the version increment in the original branch.

Ah yeah, I'll add this.

**Comment by michalborkowski96** [comment]
@adriansmares I've added a fixup commit

File: .goreleaser.release.yml
Comment (htdvisser) [reply .goreleaser.release.yml]: Because that would result in older versions getting pushed as "latest" if we did backport releases.

File: .github/workflows/create-release-branch.yml
Comment (htdvisser) [inline .github/workflows/create-release-branch.yml:21]: Do we need this or is it just for development / debugging?

File: .github/workflows/create-release-branch.yml
Comment (htdvisser) [inline .github/workflows/create-release-branch.yml]: This doesn't look formatted.

File: .goreleaser.release.yml
Comment (adriansmares) [reply .goreleaser.release.yml]: It would have been nice if we could control whenever or not this tag is pushed via the action parameters. But I don't think that's easy so probably we'll keep the manual push.

File: .github/workflows/create-release-branch.yml
Comment (htdvisser) [inline .github/workflows/create-release-branch.yml:67]: Doesn't `actions/checkout@v2` already do this?

File: .github/workflows/create-release-branch.yml
Comment (htdvisser) [inline .github/workflows/create-release-branch.yml:73]: I'd prefer explicitly adding things that we know we want to add

File: .github/workflows/create-release-branch.yml
Comment (htdvisser) [inline .github/workflows/create-release-branch.yml:10]: Nit: empty line above

File: .github/workflows/release-rc.yml
Comment (htdvisser) [inline .github/workflows/release-rc.yml:20]: Same

File: .github/workflows/release-tag.yml
Comment (htdvisser) [inline .github/workflows/release-tag.yml]: `github.repository_owner` here is `TheThingsNetwork`, and you can't assign issues to an organization afaik.

File: .github/workflows/release-tag.yml
Comment (htdvisser) [inline .github/workflows/release-tag.yml]: Same

File: .goreleaser.release.yml
Comment (htdvisser) [reply .goreleaser.release.yml]: It is also possible to determine that in the GitHub Actions workflow. If we tag a backport, we'd do so for a previous minor version, so if we just look at the existing tags in the repository, we can only perform that step if we're `>=` the latest tag.

File: .github/workflows/create-release-branch.yml
Comment (michalborkowski96) [reply .github/workflows/create-release-branch.yml:21]: We do need this to make sure that this action is only run on a `v3.x` branch. We don't want things like "create a release branch based on a `feature/something` branch"

File: .github/workflows/create-release-branch.yml
Comment (michalborkowski96) [reply .github/workflows/create-release-branch.yml]: What do you mean?

File: .github/workflows/create-release-branch.yml
Comment (michalborkowski96) [reply .github/workflows/create-release-branch.yml:67]: Why would it?

File: .github/workflows/release-tag.yml
Comment (michalborkowski96) [reply .github/workflows/release-tag.yml]: Yeah, I didn't know what to do here. We want to assign this PR to someone, but it's not clear to me who would be the assignee.

Comment (adriansmares) [reply .github/workflows/release-tag.yml]: Perhaps assign it to a team ? `@TheThingsNetwork/stack-maintainers`

File: .github/workflows/create-release-branch.yml
Comment (michalborkowski96) [reply .github/workflows/create-release-branch.yml:73]: From design perspective it's kinda difficult, because we delegate certain things to commands. "mage, please update version files, I don't care how you do it, just do it".

Since we always execute the same few commands, it's not like much can go wrong.

**Comment by michalborkowski96** [comment]
@adriansmares @htdvisser added another fixup

File: .github/workflows/create-release-branch.yml
Comment (htdvisser) [reply .github/workflows/create-release-branch.yml]: I mean that it looks like your format-on-save isn't working (at least, for yml files).

```suggestion
  RELEASE_VERSION: "${{ format('{0}.{1}', github.ref_name, github.event.inputs.patch_version) }}"
```

File: .github/workflows/create-release-branch.yml
Comment (htdvisser) [reply .github/workflows/create-release-branch.yml:67]: nvm, I overlooked the `--remote`.

File: .github/workflows/release-tag.yml
Comment (htdvisser) [inline .github/workflows/release-tag.yml]: Have you tested this? Because as far as I know, you can't assign PRs to teams (yet; would be nice to have).

File: .github/workflows/release-rc.yml
Comment (htdvisser) [inline .github/workflows/release-rc.yml:151]: In the past we also pushed `latest` to GHCR (though not consistently; it's currently on 3.16.2), so perhaps we should also push to GHCR here?

Comment (michalborkowski96) [reply .github/workflows/release-rc.yml:151]: We can fill a separate issue for things that aren't included in the current release flow but would be nice to have

File: .github/workflows/release-tag.yml
Comment (michalborkowski96) [reply .github/workflows/release-tag.yml]: No. First I want to go through all the review rounds, and at the end test. It doesn't make sense to go through entire workflow after every single comment

File: .github/workflows/release-rc.yml
Comment (htdvisser) [reply .github/workflows/release-rc.yml:151]: Fine with me.

File: .github/workflows/release-tag.yml
Comment (adriansmares) [reply .github/workflows/release-tag.yml]: FTR my comments have been dealt with, so feel free to test this.

**Comment by michalborkowski96** [comment]
I've done tests on my fork, and looks like things should be fine

File: .github/workflows/create-release-branch.yml
Comment (adriansmares) [reply .github/workflows/create-release-branch.yml]: Don't forget to fix this before merging.

**Review by adriansmares — APPROVED** [review]

File: .github/workflows/create-release-branch.yml
Comment (htdvisser) [reply .github/workflows/create-release-branch.yml:73]: I don't know. The way I see it, we know what _should_ change, so we can just explicitly add those files, being confident that we're not committing unexpected changes.

@adriansmares what do you think?

File: .github/workflows/create-release-branch.yml
Comment (htdvisser) [reply .github/workflows/create-release-branch.yml]: Ideally these kind of things should be addressed (or discussed) _before_ requesting another review, so that we don't waste time requesting the same changes again.

File: .github/workflows/create-release-branch.yml
Comment (adriansmares) [reply .github/workflows/create-release-branch.yml:73]: This flow can generate only release candidates, and the promotion itself is a manual process, so it should not be the case that releasing changes which have not been reviewed.

We can rework this step to be a PR instead of a commit to the release branch, but do we need that ? The promotion itself is manual.

**Review by htdvisser — APPROVED** [review]
You didn't have to format _all_ YAML files, I only asked you to fix the formatting of the ones where you made changes. But fine.

**Pull request merged**
